pkg/agent/iptables/iptables.go

```diff
@@ -118,6 +156,126 @@ func (c *Client) SetupRules() error {
 		if err := c.ensureRule(rule.table, rule.chain, ruleSpec); err != nil {
 			return err
 		}
+		c.storeRule(rule.table, rule.chain, ruleSpec)
+	}
+	return nil
+}
+
+// AddPeerCIDR adds iptable rules relevant to peerPodCIDR
+// It's idempotent and can be safely called on every startup.
+func (c *Client) AddPeerCIDR(peerPodCIDR *net.IPNet, peerNodeIP net.IP) error {
+	if !c.nodeConfig.TrafficEncapMode.NeedEncap(peerNodeIP, c.nodeConfig.NodeIPAddr) {
+		ruleSpec := []string{
+			"-i", c.hostGateway, "!", "-o", c.hostGateway,
+			"-d", peerPodCIDR.String(), "-s", c.nodeConfig.PodCIDR.String(), "-j", AcceptTarget,
```

Comment:
that will be the best if agent knows cluster CIDR, not sure what that info comes from though?

Reply:
One way is to require user to provide it, like ServiceCIDR. @tnqn: do you know any better way?

Reply:
Is it necessary to limit the source IP must be local Pod CIDR (`"-s", c.nodeConfig.PodCIDR.String()`)? Will it block external Service traffic, like NodePort?

Reply:
we can perhaps further limit traffic with -s podCIDR, but it may not be required because anything comes out of gw0 must come from local pod, 

traffic to service IP will not has dstIP in peerPodCIDR.

Reply:
I guess I might not present my question clearly. Let's think about this senario, an external address is accessing a NodePort Service. The source IP should be the external address, and I am wondering if the packet could be dropped by this iptables rule, for the source address is not in "c.nodeConfig.PodCIDR" or "podCIDR".

Besides,  the traffic to service IP is using Service IP as the destination, but after the kube-proxy works, the destination IP should be changed to the chosen backend Pod IP. If the Pod is on peer Node, then the destionation IP should be in peerPodCIDR.

Reply:
1. if this is external traffic for nodePort service, this rule does not apply because traffic does not satisfy inport=gw0 and outport!=gw0

2. the reason this rule exists if so that pod-to-pod traffic does not get masqueraded, and that is all. please see comment.

Reply:
> One way is to require user to provide it, like ServiceCIDR. @tnqn: do you know any better way?

AFAIK, we can only get it via a config provided by user. Even kube-proxy which needs it can only take it from config and will sacrifice some feature if not provided:
https://github.com/kubernetes/kubernetes/blob/641321c94c3e0943a57230596164e2e8f2eb2c52/cmd/kube-proxy/app/server.go#L148
https://github.com/kubernetes/kubernetes/blob/641321c94c3e0943a57230596164e2e8f2eb2c52/pkg/proxy/iptables/proxier.go#L288